### PR TITLE
o/configstate: deal with no longer valid refresh.timer=managed

### PR DESCRIPTION
**Pull request by stolowski**
Deal with no longer valid refresh.timer=managed / refresh.schedule=managed in the state when validating refresh config option of core. This option may become invalid if the snap managing refreshes is not signed anymore (i.e. devicestate.CanManageRefreshes returns false). The idea is to detect whether the value came in the transaction changes or became invalid in the existing config; if it's the latter, then it's ignored and doesn't interrupt the processing of remaining options. refreshScheduleManaged() handles this correctly and if "managed" cannot be supported anymore, we log it.

This isn't ideal. Ideally config infra would handle this for all options (but this probably needs discussion and is a larger piece of work). If this fix seems acceptable, then I'll try to work on a spread test.

Related bug: https://bugs.launchpad.net/snapd/+bug/1899992

### CODE REVIEW

File: overlord/configstate/configcore/refresh.go [hunk 1]
```diff
@@ -40,6 +40,19 @@ func init() {
...
 }
 
+func reportOrResetInvalidManageRefreshes(tr config.Conf, optName string) error {
+	// check if the option is set as part of transaction changes; if not than
+	// it's already set in the config state.
+	for _, k := range tr.Changes() {
+		if k == "core."+optName {
+			return fmt.Errorf("cannot set schedule to managed")
+		}
+	}
```
Comment (pedronis) [inline overlord/configstate/configcore/refresh.go:50]: given that afaik we don't have at any other option that can be invalidated after being set? (is this true?) doing something only here seems ok

File: overlord/configstate/configcore/refresh.go [hunk 2]
```diff
@@ -40,6 +40,19 @@ func init() {
...
+	tr.Set("core", optName, nil)
```
Comment (pedronis) [inline overlord/configstate/configcore/refresh.go:51]: I'm a bit unsure whethere resetting the option is what we want vs just ignoring it though. The code in refreshScheduleManaged in autorefresh.go will do anyway the right thing ignoring the setting?

File: overlord/configstate/configcore/refresh.go [hunk 3]
```diff
@@ -40,6 +40,19 @@ func init() {
...
+	tr.State().Warnf(`core config option %s="managed" cannot be used with current snap setup and has been reset`, optName)
```
Comment (pedronis) [inline overlord/configstate/configcore/refresh.go:52]: I understand that the user is at the terminal at this point otoh is a bit strange to mention an option that is not getting touched here,   maybe this warning needs to be in autorefresh.go inside/near  refreshScheduleManaged perhaps?

File: overlord/configstate/configcore/refresh.go (hunk 1 above)
Comment (pedronis) [reply overlord/configstate/configcore/refresh.go:50]: we probably want to expand the comment a bit though but how depends a bit on what we decided to do about my other points

**Review by pedronis — COMMENTED** [review]
basic idea looks fine, but some wondering about the resetting and warning

File: overlord/configstate/configcore/refresh.go (hunk 2 above)
Comment (stolowski) [reply overlord/configstate/configcore/refresh.go:51]: You're right; we logger.Noticef in refreshScheduleManaged. Whether this should be turned into a warning there is an open question I guess?

File: overlord/configstate/configcore/refresh.go (hunk 1 above)
Comment (stolowski) [reply overlord/configstate/configcore/refresh.go:52]: Done.

**Review by pedronis — APPROVED** [review]
thank you

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for this

**Comment by stolowski** [comment]
google:ubuntu-20.04-64:tests/main/interfaces-snapd-control-with-manage fails for real (may need updating), investigating

**Review by mvo5 — APPROVED** [review]
Thanks

**Pull request merged**